### PR TITLE
add proxy methods for MGI stack

### PR DESCRIPTION
**Pull request by deanshackley**

### CODE REVIEW

File: assets/lambda/cancelledTradeProxy.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  detail: CancelExecutedMgiTrade
+) => {
+  return {
+    detail: detail,
+    getDestinationDetailType: () => `cancel-${detail.trading_model}-trade`,
+    source: 'CancelMgiExecutedTrade',
+  };
+};
+
+export const handler = (event: APIGatewayProxyEvent) => {
```
Comment (eoln) [inline assets/lambda/cancelledTradeProxy.ts:20]: for arrow function we can use shortcut notation without `return` keyword, let drop passing `log` as param.
```suggestion
export const handler = (event: APIGatewayProxyEvent) => handleEvent<CancelExecutedMgiTrade>(event, factory);
```

File: assets/lambda/cancelledTradeProxy.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  return handleEvent<CancelExecutedMgiTrade>(event, factory, log);
```
Comment (eoln) [inline assets/lambda/cancelledTradeProxy.ts:21]:
```suggestion
```

File: assets/lambda/cancelledTradeProxy.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+};
```
Comment (eoln) [inline assets/lambda/cancelledTradeProxy.ts:22]:
```suggestion
```

File: assets/lambda/lib/eventHandler.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
+const logClass = require('lambda-log');
+import * as eb from '@aws-sdk/client-eventbridge';
+import {APIGatewayProxyEvent} from 'aws-lambda';
+
+const {SYSTEM_EVENT_BUS_NAME, MTFX_AWS_REGION} = process.env;
+
+const ebClient = new eb.EventBridgeClient({region: MTFX_AWS_REGION});
+
+export interface IProxyEvent<T> {
```
Comment (eoln) [inline assets/lambda/lib/eventHandler.ts:9]: in other parts of code intefaces are declared without `I` prefix - let keep one style.
```suggestion
export interface ProxyEvent<T> {
```

File: assets/lambda/lib/eventHandler.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  detail: T;
+  getDestinationDetailType(): string;
+  source: string;
+}
+
+export type ProxyFactory<T> = (type: T) => IProxyEvent<T>;
```
Comment (eoln) [inline assets/lambda/lib/eventHandler.ts:15]:
```suggestion
export type ProxyFactory<T> = (type: T) => ProxyEvent<T>;
```

File: assets/lambda/lib/eventHandler.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+export async function handleEvent<T>(
+  event: APIGatewayProxyEvent,
+  factory: ProxyFactory<T>,
```
Comment (eoln) [inline assets/lambda/lib/eventHandler.ts:19]:
```suggestion
  factory: ProxyFactory<T>
```

File: assets/lambda/orderStatusProxy.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+const log = require('lambda-log');
```
Comment (eoln) [inline assets/lambda/orderStatusProxy.ts:1]:
```suggestion
import * as log from 'lambda-log';
```

File: assets/lambda/orderStatusProxy.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+const factory: ProxyFactory<ProxyMgiStatus> = (detail: ProxyMgiStatus) => {
+  return {
+    detail: detail,
+    getDestinationDetailType: () => 'update-order-status',
+    source: 'ProxyMgiStatus',
+  };
+};
+
+export const handler = (event: APIGatewayProxyEvent) => {
```
Comment (eoln) [inline assets/lambda/orderStatusProxy.ts:18]:
```suggestion
export const handler = (event: APIGatewayProxyEvent) => handleEvent<ProxyMgiStatus>(event, factory);
```

File: assets/lambda/orderStatusProxy.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  return handleEvent<ProxyMgiStatus>(event, factory, log);
```
Comment (eoln) [inline assets/lambda/orderStatusProxy.ts:19]:
```suggestion
```

File: assets/lambda/orderStatusProxy.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+};
```
Comment (eoln) [inline assets/lambda/orderStatusProxy.ts:20]:
```suggestion
```

File: lib/MgiProxyStack.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+        authorizationType: apigw.AuthorizationType.IAM,
+      }
+    );
+  }
+
+  createLayer(): lambda.LayerVersion {
+    const layerFiles = path.resolve(__dirname, '../assets/layer');
+    const lambdaLayer = new lambda.LayerVersion(this, 'mgiProxyLayer', {
+      code: lambda.Code.fromAsset(layerFiles),
+      compatibleRuntimes: [lambda.Runtime.NODEJS_12_X],
```
Comment (eoln) [inline lib/MgiProxyStack.ts:93]: Let use node runtime as fresh as possible
```suggestion
      compatibleRuntimes: [lambda.Runtime.NODEJS_14_X],
```

File: lib/MgiProxyStack.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        ResourceType.LAMBDA_LAYER
+      ),
+    });
+
+    return lambdaLayer;
+  }
+
+  createLambda(name: string, systemEventBusName: string) {
+    return new lambda.Function(this, `${name}Lambda`, {
+      runtime: lambda.Runtime.NODEJS_12_X,
```
Comment (eoln) [inline lib/MgiProxyStack.ts:105]:
```suggestion
      runtime: lambda.Runtime.NODEJS_14_X,
```

File: package.json [hunk 13]
```diff
@@ -18,30 +18,32 @@
     "init": "node init.js"
   },
   "devDependencies": {
-    "@aws-cdk/assert": "1.106.1",
-    "@milltechfx/base-aws-stack": "^1.0.56",
+    "@aws-cdk/assert": "^1.111.0",
```
Comment (eoln) [inline package.json:21]: for importing any `@aws-cdk/*` dependencies let use fixed version - elsewhere we will have problems...
```suggestion
    "@aws-cdk/assert": "1.111.0",
```

File: package.json [hunk 14]
```diff
@@ -18,30 +18,32 @@
...
+    "@milltechfx/base-aws-stack": "^1.0.66",
     "@milltechfx/common": "^1.1.0",
     "@types/aws-lambda": "^8.10.76",
     "@types/jest": "^26.0.23",
     "@types/node": "^15.6.1",
-    "aws-cdk": "1.106.1",
+    "aws-cdk": "^1.111.0",
```
Comment (eoln) [inline package.json:27]: I would suggest to specify only deps  in the way `@aws-cdk/<functionality>` for constructs used to create the stack instead depend on everything
```suggestion```

File: package.json [hunk 15]
```diff
@@ -18,30 +18,32 @@
...
     "source-map-support": "^0.5.19",
-    "ts-jest": "^27.0.2",
+    "ts-jest": "^26.5.6",
     "ts-node": "^10.0.0",
     "typescript": "4.3.2"
   },
   "dependencies": {
-    "@aws-cdk/aws-lambda": "^1.106.1",
-    "@aws-cdk/core": "^1.106.1",
+    "@aws-cdk/aws-apigateway": "^1.111.0",
```
Comment (eoln) [inline package.json:37]:
```suggestion
    "@aws-cdk/aws-apigateway": "1.111.0",
```

File: package.json [hunk 16]
```diff
@@ -18,30 +18,32 @@
...
+    "@aws-cdk/aws-events": "^1.111.0",
```
Comment (eoln) [inline package.json:38]:
```suggestion
    "@aws-cdk/aws-events": "1.111.0",
```

File: package.json [hunk 17]
```diff
@@ -18,30 +18,32 @@
...
+    "@aws-cdk/aws-iam": "^1.111.0",
```
Comment (eoln) [inline package.json:39]:
```suggestion
    "@aws-cdk/aws-iam": "1.111.0",
```

File: package.json [hunk 18]
```diff
@@ -18,30 +18,32 @@
...
+    "@aws-cdk/aws-lambda": "^1.111.0",
```
Comment (eoln) [inline package.json:40]:
```suggestion
    "@aws-cdk/aws-lambda": "1.111.0",
```

File: package.json [hunk 19]
```diff
@@ -18,30 +18,32 @@
...
+    "@aws-cdk/core": "^1.111.0",
```
Comment (eoln) [inline package.json:41]:
```suggestion
    "@aws-cdk/core": "1.111.0",
```

File: test/MgiProxy.test.ts [hunk 20]
```diff
@@ -0,0 +1,13 @@
+process.env.MTFX_AWS_REGION = 'eu-west-1';
+
+import * as cdk from '@aws-cdk/core';
+import * as MgiProxy from '../lib/MgiProxyStack';
+
+test('Empty Stack', () => {
```
Comment (eoln) [inline test/MgiProxy.test.ts:6]: let use `describe` as a wrapper

File: test/cancelledTradeProxy.test.ts [hunk 21]
```diff
@@ -0,0 +1,74 @@
+process.env.SYSTEM_EVENT_BUS_NAME = 'mytesteventbus';
+import {mockClient} from 'aws-sdk-client-mock';
+import {handler} from '../assets/lambda/cancelledTradeProxy';
+import {createEvent} from './fakeApiGatewayEvent';
+import {EventBridgeClient, PutEventsCommand} from '@aws-sdk/client-eventbridge';
+const eventBridgeMock = mockClient(EventBridgeClient);
+
```
Comment (eoln) [inline test/cancelledTradeProxy.test.ts:7]: let use describe as a wrapper.

File: test/cancelledTradeProxy.test.ts [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+beforeEach(() => {
+  eventBridgeMock.on(PutEventsCommand).resolves({
```
Comment (eoln) [inline test/cancelledTradeProxy.test.ts:9]: there is a need to also test the case when `FailedEntryCount` is bigger then `0`, or prop is not present in `putResult`

File: test/cancelledTradeProxy.test.ts [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+
+test('Should raise event for agency cancel', async () => {
+  const body = {
+    order_id: '111',
+    mgi_version: '12',
+    mgi_last_moditifed_time: '2021-8-16',
+    mgi_cancellation_reason: 'rebook',
+    trading_model: 'agency',
+  };
+
```
Comment (eoln) [inline test/cancelledTradeProxy.test.ts:26]: let be sure what that execution environment has proper state before making a call.
```suggestion
expect(eventBridgeMock.calls()).toHaveLength(0);
```

File: test/fakeApiGatewayEvent.ts [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+  const bodyText = JSON.stringify(body);
+
+  const evt: APIGatewayProxyEvent = {
+    body: bodyText,
+    headers: {},
+    multiValueHeaders: {},
+    httpMethod: 'POST',
+    isBase64Encoded: false,
+    path: 'ProxyMgiStatus',
+    requestContext: request,
```
Comment (eoln) [inline test/fakeApiGatewayEvent.ts:61]: use type casting to have minimal mocking
```suggestion
    requestContext: {} as unknown as APIGatewayEventRequestContext,
```

File: test/fakeApiGatewayEvent.ts [hunk 25]
```diff
@@ -0,0 +1,70 @@
+import {
+  APIGatewayProxyEvent,
+  APIGatewayEventRequestContext,
+  APIGatewayEventDefaultAuthorizerContext,
+  APIGatewayEventIdentity,
+} from 'aws-lambda';
+
+export function createEvent(body:any): APIGatewayProxyEvent {
+  const auth: APIGatewayEventDefaultAuthorizerContext = {};
+  const identity: APIGatewayEventIdentity = {
```
Comment (eoln) [inline test/fakeApiGatewayEvent.ts:10]: there is no need to mock this elements in every detail, our code is not using these properties

File: test/fakeApiGatewayEvent.ts [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+    cognitoIdentityId: '',
+    cognitoIdentityPoolId: '',
+    principalOrgId: '',
+    sourceIp: '',
+    user: '',
+    userAgent: '',
+    userArn: '',
+  };
+
+  const request: APIGatewayEventRequestContext = {
```
Comment (eoln) [inline test/fakeApiGatewayEvent.ts:27]: no need to mock this request in detail

File: test/orderStatusProxy.test.ts [hunk 27]
```diff
@@ -0,0 +1,44 @@
+process.env.SYSTEM_EVENT_BUS_NAME = 'mytesteventbus';
+import {mockClient} from 'aws-sdk-client-mock';
+import {handler} from '../assets/lambda/orderStatusProxy';
+import {createEvent} from './fakeApiGatewayEvent';
+import {EventBridgeClient, PutEventsCommand} from '@aws-sdk/client-eventbridge';
+const eventBridgeMock = mockClient(EventBridgeClient);
+
+beforeEach(() => {
+  eventBridgeMock.on(PutEventsCommand).resolves({
```
Comment (eoln) [inline test/orderStatusProxy.test.ts:9]: there is a need to implement tests when `FailedEntryCount` is bigger than `0` or is not present in `putResponse`

**Review by eoln — CHANGES_REQUESTED** [review]
Great work done Dean! Wellcome in Typescript wilderness. Nice to see someone is using generic types! I like your clean code!

Although the here is a need to make some stylistic changes and improve tests

**Comment by eoln** [comment]
`test/MgiProxy.test.ts` is failing in Teamcity -> you need to build assets before running stack related tests

File: assets/lambda/lib/eventHandler.ts [hunk 28]
```diff
@@ -0,0 +1,92 @@
+const logClass = require('lambda-log');
```
[the remark this answers is not in the document]
Comment (deanshackley) [reply assets/lambda/lib/eventHandler.ts]: done

File: assets/lambda/cancelledTradeProxy.ts [hunk 29]
```diff
@@ -0,0 +1,22 @@
+const log = require('lambda-log');
```
[the remark this answers is not in the document]
Comment (deanshackley) [reply assets/lambda/cancelledTradeProxy.ts]: done

File: assets/lambda/lib/eventHandler.ts [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+  getDestinationDetailType(): string;
+  source: string;
+}
+
+export type ProxyFactory<T> = (type: T) => IProxyEvent<T>;
+
+export async function handleEvent<T>(
+  event: APIGatewayProxyEvent,
+  factory: ProxyFactory<T>,
+  log: typeof logClass
```
[the remark this answers is not in the document]
Comment (deanshackley) [reply assets/lambda/lib/eventHandler.ts]: done

File: assets/lambda/lib/eventHandler.ts [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+) {
+  log.options.meta.event = event;
+  log.info('Event received');
+
+  try {
+    const body = JSON.parse(event.body || '');
```
[the remark this answers is not in the document]
Comment (deanshackley) [reply assets/lambda/lib/eventHandler.ts]: done

File: assets/lambda/lib/eventHandler.ts [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+    const detail: T = body;
```
[the remark this answers is not in the document]
Comment (deanshackley) [reply assets/lambda/lib/eventHandler.ts]: done

File: test/cancelledTradeProxy.test.ts (hunk 23 above)
Comment (deanshackley) [reply test/cancelledTradeProxy.test.ts:18]: done

File: test/fakeApiGatewayEvent.ts (hunk 24 above)
Comment (deanshackley) [reply test/fakeApiGatewayEvent.ts]: done, great, great tip, thanks

File: test/fakeApiGatewayEvent.ts (hunk 25 above)
Comment (deanshackley) [reply test/fakeApiGatewayEvent.ts]: done

File: test/fakeApiGatewayEvent.ts (hunk 26 above)
Comment (deanshackley) [reply test/fakeApiGatewayEvent.ts]: done

File: test/orderStatusProxy.test.ts (hunk 27 above)
Comment (deanshackley) [reply test/orderStatusProxy.test.ts]: done

**Review by eoln — APPROVED** [review]
lgtm

**Pull request closed without merging**